### ISSUE TITLE
[DESIGN] "Configure Plugins" screen redesign.

### ISSUE THREAD

**Issue by me4oslav** [fossfreedom/alternative-toolbar#37]
Before I start on the actual design rationale I must say that I am totally unsure if this screen can be changed via a plugin and even if that's possible, if this plugin is the correct one to file a bug report to.
Now, having said that - I have noticed that the "Configure plugins" has really outdated design that looks bad and also breaks the Gnome HIG. I did a quick redesign of it that looks better and obeys the Gnome HIG:

![plugins screen](https://cloud.githubusercontent.com/assets/1658870/7366485/56c64a3c-eda1-11e4-8223-22544e521892.png)

Rationale:
If possible it should become a screen with a headerbar (perhaps even modal dialog if possible) and close button on the right of it.
You have the plugins names followed by plugins description below them on the left and on/off switches for enabling/disabling a plugin on the right.
Below the plugins pane you have inline toolbar with a button <preferences | about> on the right and if doable <remove> button on the right. The symbolics for the button on the right are standard ones - "preferences-system-symbolic" and "dialog-information-symbolic".

P.S.
I don't know how much of this is doable (for example the CSD and the remove button don't exists as functions in the current implementation), so I know that if implementation is doable it might not be possible to match the wireframe perfectly. I was also considering requesting "Search", but decided that this is bit over the top. This design is a common thing in Gnome as is in System Settings for example:

![screenshot from 2015-04-28 12 12 17](https://cloud.githubusercontent.com/assets/1658870/7366281/d5229fe0-ed9f-11e4-8183-2d5f8ca5c9bb.png)

So, can this be implemented and if so, how much of it?

**Comment by fossfreedom** [comment]
In terms of the rhythmbox API - there are no direct hooks to do this.

In terms of GTK - it should be possible to redirect the Gio.Action which is probably called when the plugins menu option is chosen to a custom implementation.

Just like the side-bar implementation though - this would be effectively a complete reimplementation of this area from ground up.  So yes - with time - it should be possible to do this.

In terms of efficiency - it would be easier and quicker to make the GUI changes directly in the RB source ... assuming you have a GTK and C programmer that would want to produce a patch.  Given my limited knowledge of the source, this is probably a glade file anyway - so redesigning the front-end should be relatively simple.

If you took the latter route ... would this be accepted upstream?  Maybe.  Looking at the last time a dev wanted to implement headerbars in various dialogs such as the Song Dialog, Jonathan (the RB maintainer) wasnt particularly keen.

 - https://bugzilla.gnome.org/show_bug.cgi?id=742224

Basically it came down to Jonathan not wanting to up the minimum GTK version requirement of rhythmbox (GTK v3.8).  Headerbars came in at GTK3.10 - but most of the nice stuff came in at GTK3.12 and GTK 3.14..  I dont know if Jonathan will up the min GTK version ... but I noted in another bug-report that he himself has inadvertently made some parts of RB GTK 3.12 anyway with some of the recent commits.  So maybe by the end of the year, RB would have been moved to GTK 3.12 or later.

Thus it comes down to priorities.  Is the plugin dialog really a high enough priority to investigate and spend time on, or are there better areas that users interact with on a day-to-day basis that developers should concentrate their efforts on.

**Comment by fossfreedom** [comment]
note - whilst thinking through this I may have missed an important fact.  This plugin dialog is probably part of LibPeas - GEdit uses the same library so probably displays the same dialog unless someone has written a custom screen.

**Comment by me4oslav** [comment]
Yes, it is a common open source practice to take the "one size fits all" policy - making an app work in bunch of desktop environment, but in the process sacrificing the design quality of the app. Gnome has one HIG, Elementary has another, KDE has a third one, XFCE - fourth and so on and so fort, so trying to create an app that works on all of these results in app that fits neither. The expression "jack of all trades, but master of none" comes to my mind.
I would rather have X apps created specifically for each X number DE, where 1 one fits only one DE, rather than software that works on all X DEs, which has worse design qualityat than specifically crafted apps. One of the reason open source design is bad is this "one size fits all policy" and when Gnome and Elementary (for example) started creating specifically tailored apps their design became much better.
Here is what I think - last time we/I tried to propose an upstream change to Rhythmbox we/I hit a bummer - the sidebar symbolic icons situation with the comment that the symbolic that was literally a radio didn't imply radio enough, so I am more or less not that keen to propose suggestions upstream.
However, what do you think? Would you try your luck upstream first or would you implement this in this plugin? 
P.S. This and the plugin's preferences dialog redesign are the last two major design changes related issues that I am filling for this plugin for a while, because after those two and the remaining bugs are covered, the plugin should be done design wise for the time being. Later on I might suggest other things, but I wanna focus on coverart plugin's design.

**Comment by fossfreedom** [comment]
I'll have a look later to see where this dialog is coded - if its part of Rhythmbox - or part of LibPeas.

If its LibPeas - then that crowd maybe more receptive with this idea.

I'll also look later to see if it is indeed possible to intercept the menu call and look more closely at the LibPeas API to see what is actually possible.  Once done I can then give you a more informed view of the proposal.

**Comment by me4oslav** [comment]
Great. I am off to redesign this plugin's preferences. Shouldn't take much time.

**Comment by fossfreedom** [comment]
In terms of the preferences - similarly - need to have a look at if we can intercept the menu call.  Once confirmed then in reality, the preferences screen basically hooks up to the various GSettings.

So - if you yourself (or someone you think can do so) can create via glade how you want the screen to look like - I can just hook up each checkbox/radio button etc with the appropriate GSetting itself.

I'll get back to you with the first point ... can the menu action be intercepted to display the new dialog.

**Comment by me4oslav** [comment]
I think i can do the glade work. Even with Glade being less then ideal tool, it still shouldn't be that hard.

**Comment by fossfreedom** [comment]
wait ... are you referring to rhythmbox preferences screen - or specifically about the alternative-toolbar plugin preferences dialog?

If the latter- the glade file is in the subfolder ui/altpreferences.ui - feel free to fiddle with this.

**Comment by me4oslav** [comment]
The latter. This one requires dynamic implementation, which is beyond my capabilities.

**Comment by fossfreedom** [comment]
ok - just shout with your ideas - lets see what we can do.

**Comment by fossfreedom** [comment]
Re - the plugins window.  Can confirm that this is indeed libpeas - rhythmbox just calls the libpeas routine to display the plugins window

 - https://developer.gnome.org/libpeas/stable/PeasGtkPluginManager.html

Having looked at the API, there doesn't appear to be a way to rework this interface - there isn't the API hooks and signals to sufficiently plug into your own code.  There doesn't appear to be a way to go from the plugins tree view to each plugin preference screen.

Thus - not possible to rework this area via a plugin.  It is indeed a separate group of devs working in this area - the library of code is used across all Gnome apps that have plugin capabilities - rhythmbox/totem/gedit

 - suggest contact these Gnome guys/girls via gnome bugzilla

**Comment by me4oslav** [comment]
So, I guess we should file a bug report here, right? https://bugzilla.gnome.org/buglist.cgi?bug_status=UNCONFIRMED&bug_status=NEW&bug_status=ASSIGNED&bug_status=REOPENED&component=libpeas-gtk&list_id=26379&product=libpeas

I also know a few Gnome designers, so I can ask them too.

**Comment by fossfreedom** [comment]
yep - looks like there is an existing bug report where various discussions about this dialog is already occurring

 - https://bugzilla.gnome.org/show_bug.cgi?id=728025

**Comment by me4oslav** [comment]
Just commented there.

**Comment by fossfreedom** [comment]
ok,

  below is not going to make much sense to most people ... but to summarise, I've had a relook at the API.  I may have found a way to possibly implement this.  I'll leave the rhythmbox python-console trace below for me (or someone else) to look at.

<pre>
You can access the main window through the 'shell' variable :
<Shell object at 0x7f800467a168 (RBShell at 0x26e1210)>
>>> from gi.repository import Peas
>>> peas = Peas.Engine.get_default()
>>> loaded = peas.get_loaded_plugins)
>>> loaded = peas.get_loaded_plugins()
>>> loaded
['soundcloud', 'smallwindow', 'notification', 'iradio', 'mpris', 'generic-player', 'power-manager', 'dbus-media-server', 'equalizer', 'audiocd', 'mmkeys', 'pythonconsole', 'spectrum', 'audioscrobbler', 'coverart_search_providers', 'mtpdevice', 'fmradio', 'rb', 'coverart_browser', 'alternative-toolbar']
>>> peas.get_plugin_info('alternative-toolbar')
<PeasPluginInfo at 0x2ed1e70>
>>> info = peas.get_plugin_info('alternative-toolbar')
>>> from gi.repository import PeasGtk
>>> ext = peas.create_extension(info, PeasGtk.Configurable, None)
>>> print (ext)
<__main__.PeasExtensionPython+PeasGtkConfigurable object at 0x7f8004011b88 (PeasExtensionPython+PeasGtkConfigurable at 0x4a8b480)>
>>> widget = ext.create_configure_widget()
DEBUG - create_display_contents
looking for ui/altpreferences.ui in /home/foss/.local/share/rhythmbox/plugins/alternative-toolbar
>>> print (widget)
<Box object at 0x7f8004011f30 (GtkBox at 0x471e450)>
>>> from gi.repository import Gtk
>>> dlg = Gtk.Dialog.new()
>>> dlg.add_button("Close", Gtk.ResponseType.CLOSE)
<Button object at 0x7f8004011d80 (GtkButton at 0x47157d0)>
>>> box = dlg.get_content_area()
>>> box.pack_start(widget, True, True, 0)
>>> dlg.show_all()
</pre>

**Comment by me4oslav** [comment]
No response from bugzilla yet.
I will try to bug some Gnome designers.

**Comment by fossfreedom** [comment]
Ok - sketched something out for this on the plugins branch.  Have a play.  Let me know what you think.

**Comment by me4oslav** [comment]
Right, will test tomorrow and tell you.

**Comment by me4oslav** [comment]
Ok, I noticed a few issues:
  -  The default window size should be bigger. Something like this should be good enough:
![screenshot from 2015-05-06 21 11 03](https://cloud.githubusercontent.com/assets/1658870/7510536/bab500be-f4a5-11e4-9e1b-a36de2145439.png)
  -  I think the preferences button can be the only button on the left of the inline toolbar and info and help should go on the right end of it and become linked buttons;
  -  The help icon should be "dialog-question-symbolic" or "help-browser-symbolic";
  -  The branch seems to be a few commits behind master;

**Comment by fossfreedom** [comment]
moved the buttons around and updated the commits to be inline with master

The dialog size was set to 600 px wide - I've up'd this to 750.  However, I'm wary of making this even larger because those people running 800x600 would then be affected.

When you get some time let me know what your thoughts are for the about screen.

**Comment by me4oslav** [comment]
Ok, tested the latest version and are you sure that the about and help buttons are linked?
They look like this:
![screenshot from 2015-05-07 15 08 40](https://cloud.githubusercontent.com/assets/1658870/7514942/1b830c5e-f4cb-11e4-9092-18262691e6fa.png)
Whilst linked ones look like this:
![screenshot from 2015-05-07 15 09 40](https://cloud.githubusercontent.com/assets/1658870/7514947/24632ff2-f4cb-11e4-8d1d-3550e715a8b0.png)
Regarding the size, something like X = from 750 to 790, Y = from 460 to 480 should be good enough. Currently the width is 750, which is not bad, but the height is too small.

**Comment by me4oslav** [comment]
Oh, new commit about this issue. Will test and comment on the morning.

**Comment by me4oslav** [comment]
Righ, just tested the latest plugin branch. Now, comments:
  -  Do we need that plugins dialog resizable?
  -  How does it act when a plugin can not be loaded?
  -  For cases where the description is too long for the plugin width, we should show a tooltip with the complete description on hover;
  -  The plugin title and description labels can get moved a few pixels to the right;
  -  There seems to be something wonky with the linked buttons:
    -  What we have:
![screenshot from 2015-05-08 09 24 55](https://cloud.githubusercontent.com/assets/1658870/7531544/40dce648-f564-11e4-90bb-b5e6cc977799.png)
    -  What System Settings has as linked buttons:
![screenshot from 2015-05-08 09 25 06](https://cloud.githubusercontent.com/assets/1658870/7531553/501fbcd4-f564-11e4-8256-192f72d2e96b.png)

**Comment by fossfreedom** [comment]
thanks

 the link issue is a GTK bug - some sort of conflict with the inline-toolbar style.  Need to investigate what is going on.

re dialog resizable.  Happy to leave this as is.  No need to stop people changing window sizes where necessary.  I personally find that frustrating when apps stop people playing with window management stuff.

re the toolip/description too long.  I havent found a GTK API that will help here.  I'll continue digging.  Maybe not possible.  Tooltips are either on or off - we need an API call that says "has ellipsize been triggered" for the label and then intercept the tooltip signal to switch the tooltip on/off.

Re the title and description - will look to adding a "space" or similar blank column in a similar way as the sidebar was recently done.

When a plugin fails to load, the switch becomes greyed out - i.e. the user cannot toggle the setting.

**Comment by me4oslav** [comment]
Linking - ok;
Spacing - ok;
Tooltips - in fedy we use tooltips only for too long descriptions. The app is GJS, but @satya164 who developed it might hep you with this. Satya, any input?
Resizing - I am not against resizing when it is needed, that would be silly. I am wondering if it actually needed in this case.
Fails to load usecase - the switch and the labels should get greyed out. On hover one should get a tooltip that says:
"The plugin can't be enabled.
<REASON GOES HERE>."

**Comment by satya164** [comment]
This is the relevant part - https://github.com/satya164/fedy/blob/next/app.js#L413

**Comment by me4oslav** [comment]
Klaus t the rescue! :P
(Inside joke, you can safely ignore this comment)

**Comment by fossfreedom** [comment]
@satya164 - you beat me to the punch :)  I was literally just looking at the Pango API to see if that could help.  Anyway - that looks implementable.  Cheers.

@me4oslav - libpeas doesnt offer the actual reason why the plugin can't be enabled.  However we can add the symbolic next to the switch.

**Comment by me4oslav** [comment]
David, ok, symbolic 5px left of the switch, greying out both the label and the switch and possibly a tooltip that says "The plugin can't be enabled" will have to do. :)

**Comment by me4oslav** [comment]
Will test the latest "plugins" branch and comment in the morning.

**Comment by me4oslav** [comment]
How do I make a plugin not load, so I can test the failed to load a plugin mechanism?

**Comment by fossfreedom** [comment]
easiest way is to install another plugin then edit its primary module file with some random text

e.g.

    git clone https://github.com/fossfreedom/desktop-art
    cd desktop-art
    gedit desktop-art.py

add some random text at the top of the file - save and then try to enable the desktop-art plugin

**Comment by me4oslav** [comment]
Just one thing about the failed to load UX - can the "dialog-error-symbolic" be red and not greyed out?

**Comment by fossfreedom** [comment]
dont really know - by default it just takes the colour of your theme.

I tried to set the style-class of the icon-image to "error" but that just turned the whole image blue - not red

**Comment by me4oslav** [comment]
All right, leave making the icon red for later, it is not that important. :)
Lets see it with fixed linked buttons, 5px whitespace on the left and tooltps for too long descriptions.
I've also been thinking - we should provide a very simple graphic installer.Plus - you can use the design for it that I'll make (hopefully tomorrow) for all your other plugins.

**Comment by fossfreedom** [comment]
Assuming you do eventually want to install this on ozon, the packaging for this is really in your packagers hands.  

Installing the plugin so that it is the default for rhythmbox on startup (i.e. so the user doesnt have to go into the plugins screen to tick the box against alternative-toolbar) is very straightforward.

Get your packager to append the following to the file alternative-toolbar.plugin

    [RB]
    InitiallyEnabled=true

Thus for your distro just make sure the .rpm or whatever patches the plugin file and all will be well.

**Comment by fossfreedom** [comment]
Re the link buttons.

This is most definitely a bug in Gtk.  I've reproduced this is a separate dummy test program based on the example here

 - http://python-gtk-3-tutorial.readthedocs.org/en/latest/dialogs.html

The bug only affects dialogs - obviously the issue doesnt occur for standard windows.

Thus - we can "live" with this - or make this two separate buttons separated by, say 5 px

The long description tooltips is already implemented.  The dialog-error-symbolic already has a 5 px margin.

**Comment by me4oslav** [comment]
Will comment back in the morning.

**Comment by fossfreedom** [comment]
...  yes yes yes ... solved using completely different Gtk Container :)

**Comment by me4oslav** [comment]
This is unrelated to this issue comment, but the plugin's preferences window is resizable and it should not be, because all the options in it are displayed in the default size.
Now, related comment:
  -  There should be a "Enhanced plugins dialog" checkbox under "Use:" section in the plugin's preferences, so the user can revert back to the old one. It is what we do for tooltips, new sidebar and etc ... so we should follow the trend. Don't forget to add a tooltip for the checkbox;
I will open a new issue about graphic installer with a wireframe, where we can discuss it.

**Comment by fossfreedom** [comment]
ok - excellent.  All done.

I'll now merge this all to master.
